Ag readme (#143)
poanetwork/bridge-ui: README.md

## POA Bridge Overview

The POA Bridge allows users to transfer assets between two chains in the Ethereum ecosystem. It is composed of several elements which are located in different POA Network repositories. 

For a complete picture of the POA Bridge functionality, it is useful to explore each repository.

**Bridge Elements**
1. Bridge UI Application. A DApp interface to transfer tokens and coins between chains, located in this repository.
2. [Bridge Smart Contracts](https://github.com/poanetwork/poa-bridge-contracts). Solidity contracts used to manage bridge validators, collect signatures, and confirm asset relay and disposal.
3. [Bridge Oracle](https://github.com/poanetwork/token-bridge). An oracle written in NodeJS.
4. [Bridge Monitor](https://github.com/poanetwork/bridge-monitor). A tool for checking balances and unprocessed events in bridged networks.
5. [Bridge Deployment Playbooks](https://github.com/poanetwork/deployment-bridge). Manages configuration instructions for remote deployments and allows you to deploy separate bridge instances for validators.
6. [Bridge Monitor](https://github.com/poanetwork/bridge-monitor).

## Bridge UI Application

> 

Author: I may add a picture of the redesigned bridge here

Reviewer: Actually there is no much difference in UI between the previous and this version. Yes, please, add an example how the UI could look like.

    * HOME_OWNER_MULTISIG: Wallet address(es) of validators. For testing, you can use the same address for all address values in the file. 
    * FOREIGN_RPC_URL=https://kovan.infura.io/mew

5. Install and run the POA Bridge Oracle.
  * Got to the `sokol-kovan-bridge` folder and  `git clone https://github.com/poanetwork/token-bridge`
  * Follow instructions in the [POA NodeJS Oracle repo](https://github.com/poanetwork/token-bridge).

If successful, you will see bridge processes run when you issue a command. For example run `npm run watcher:signature-request`

**Example Output:**
```bash
[1539195000507] INFO (watcher-signature-request): Connected to redis
[1539195000545] INFO (watcher-signature-request): Connected to amqp Broker
[1539195006085] INFO (watcher-signature-request): Found 0 UserRequestForSignature events
[1539195011467] INFO (watcher-signature-request): Found 0 UserRequestForSignature events
```

6. Keep the bridge processes running. Open a separate terminal window and go to the `sokol-kovan-bridge` folder to install and unpack this repository.

  *  `git clone https://github.com/poanetwork/bridge-ui.git`  
  * `cd bridge-ui`  

>   * `npm install`  

Author: I am currently unable to npm install, there is a recurring loop.

  * Create a .env file from the example file[.env.example](.env.example)  
`cp .env.example .env`  
  * Insert addresses from the bridgeDeploymentResults.json file into the .env file.
`cat ../poa-bridge-contracts/deploy/bridgeDeploymentResults.json`  

```bash
# HomeBridge address in bridgeDeploymentResults.json
REACT_APP_HOME_BRIDGE_ADDRESS=0x.. 
#ForeignBridge address in bridgeDeploymentResults.json
REACT_APP_FOREIGN_BRIDGE_ADDRESS=0x..
#https public RPC node for Foreign network
REACT_APP_FOREIGN_HTTP_PARITY_URL=https://kovan.infura.io/mew
#public RPC node for Home network 
REACT_APP_HOME_HTTP_PARITY_URL=https://sokol.poa.network 
#// Gas price speed option (slow, standard, fast, instant)
REACT_APP_GAS_PRICE_SPEED_TYPE=fast
```

  * Run `npm run start`
  * Make sure you have a web3 wallet installed (Nifty Wallet or MetaMask) and connected to the POA Sokol Network. Check that you are connected the funded account from step 2 in the web3 wallet
  * Specify an amount and click Transfer to make a cross chain transaction from Sokol to Kovan

## Testing

> `npm run test`

Author: unable to test yet

Author: npm run test failed

The current implementation allows for several bridge modes.

  1. `Native-to-ERC20` Coins on a Home network can be converted to ERC20-compatible tokens on a Foreign network. Coins are locked on the Home side and the corresponding amount of ERC20 tokens are minted on the Foreign side. When the operation is reversed, tokens are burnt on the Foreign side and unlocked in the Home network.
  2. `ERC20-to-ERC20` ERC20-compatible tokens on the Foreign network are locked and minted as ERC20-compatible tokens (ERC677 tokens) on the Home network. When transferred from Home to Foreign, they are burnt on the Home side and unlocked in the Foreign network. This can be considered a form of atomic swap when a user swaps the token "X" in network "A" to the token "Y" in network "B".

### UI Features
- Shows daily limits in both networks
- Displays all events in both networks
- Filter events from a specific block number on both sides of the bridge
- Find a corresponding event on different sides of the bridge
- Submit a transaction from Home to Foreign network
- Submit a transaction from Foreign to Home network

### User Transactions

> - Connect to the network you want to transfer coins from using a web3 wallet such as Nifty Wallet or MetaMask. This can be the Home or Foreign network. The wallet must be funded to include amount to transfer and any gas costs.

Reviewer: 1. The suggestion is to re-phrase: `The wallet must be funded to cover any gas costs and, in case of `native-to-erc20` bridge mode, to include amount to transfer.`
2. Also, for `erc20-to-erc20` mode it makes sense to mention that account must own tokens linked with the particular bridge.

- Specify the amount to send.
- Click the Transfer button.
- Confirm the transaction via the web3 wallet. 

> - The same address is used to send a coin from the Home network and receive a token on the Foreign Network.

Reviewer: Does it make sense to add that _in order to send assets in the opposite direction it is necessary to change the network in the wallet_. It is not obvious since is not described anywhere that's why the users' questions appeared in the past.

- [Testnet Bridge URL](https://bridge-testnet.poa.net/)
- [Bridge UI Tutorial Videos](https://www.youtube.com/playlist?list=PLS5SEs8ZftgUqR3hVFiEXQLqE9QI8sIGz)
- [Article on the POA Bridge](https://medium.com/poa-network/cross-chain-bridges-paving-the-way-to-internet-of-blockchains-422ac94bc2e5)

## Getting Started

The following is an example setup using the POA Sokol testnet as the Home network, and the Ethereum Kovan testnet as the Foreign network. The instructions for the Bridge UI are identical for an `ERC20-to-ERC20` configuration, but the smart contract deployment steps will vary.

### Dependencies

- [poa-bridge-contracts](https://github.com/poanetwork/poa-bridge-contracts)
- [token-bridge](https://github.com/poanetwork/token-bridge)
- [node.js](https://nodejs.org/en/download/)
- [Nifty Wallet](https://chrome.google.com/webstore/detail/nifty-wallet/jbdaocneiiinmjbjlgalhcelgbejmnid?hl=en) or [MetaMask](https://metamask.io/)

### Example Setup

1. Create an empty folder where you will be setting up your bridge. In this example we call it `sokol-kovan-bridge`.
`mkdir sokol-kovan-bridge && cd sokol-kovan-bridge`  

> 2. Prepare temporary ETH address(es) for deployment. Each validator will require an address. 

Reviewer: You don't need a JSON keystore for validators since the Token Bridge works with the private keys. So, the step to get the validator's private key could be to export the private key from the wallet. So, the step 3 is not needed in this case.

  * Generate a new Ethereum keystore file JSON and its password.
    * Got to https://www.myetherwallet.com.
    * Enter a password and click `Create New Wallet`.
    * Download the keystore file.

3. Create a new account in Nifty Wallet or MetaMask and import the JSON keystore file. See the wallet resources if you need more information on this step.  

4. Fund the test account(s).
  * Fund Home accounts (`Validators`) using the [POA Sokol Faucet](https://faucet-sokol.herokuapp.com/)
  * Get free Kovan Coins from the [gitter channel](https://gitter.im/kovan-testnet/faucet) or [Iracus faucet](https://github.com/kovan-testnet/faucet) for Foreign Accounts. Get 5 Keth to 1 acc, and transfer from it to all other wallets.

5. Deploy the Sokol <-> Kovan Bridge contracts.
  * Go to the the `sokol-kovan-bridge` folder created in step 1 and `git clone https://github.com/poanetwork/poa-bridge-contracts`
  * Follow instructions in the [POA Bridge contracts repo](https://github.com/poanetwork/poa-bridge-contracts).
  * Set the parameters in the .env file.
    * DEPLOYMENT_ACCOUNT_PRIVATE_KEY: Private key from step 2
    * HOME_RPC_URL=https://sokol.poa.network

>     * HOME_OWNER_MULTISIG: Wallet address(es) of validators. For testing, you can use the same address for all address values in the file. 

Reviewer: Actually the same is applicable for `HOME_OWNER_MULTISIG`, `HOME_UPGRADEABLE_ADMIN_VALIDATORS`, `HOME_UPGRADEABLE_ADMIN_BRIDGE`, `FOREIGN_OWNER_MULTISIG`, `FOREIGN_UPGRADEABLE_ADMIN_VALIDATORS` and `FOREIGN_UPGRADEABLE_ADMIN_BRIDGE`

>     * FOREIGN_RPC_URL=https://kovan.infura.io/mew

Reviewer: Does it make sense to add that when the deployment is successfully finished, it necessary to make sure that `bridgeDeploymentResults.json` exists in `deploy` directory with the bridge contract addresses.

> 5. Install and run the POA Bridge Oracle.

Reviewer: Does it make sense to call it _the Token Bridge_?

  * Got to the `sokol-kovan-bridge` folder and  `git clone https://github.com/poanetwork/token-bridge`

>   * Follow instructions in the [POA NodeJS Oracle repo](https://github.com/poanetwork/token-bridge).

Reviewer: Does it make sense to call it _the Token Bridge_?

If successful, you will see bridge processes run when you issue a command. For example run `npm run watcher:signature-request`

**Example NPM Output:**
```bash
[1539195000507] INFO (watcher-signature-request): Connected to redis
[1539195000545] INFO (watcher-signature-request): Connected to amqp Broker
[1539195006085] INFO (watcher-signature-request): Found 0 UserRequestForSignature events
[1539195011467] INFO (watcher-signature-request): Found 0 UserRequestForSignature events
```

**Example Docker Output:**

> ```bash

Reviewer: I would say that you will see this depending on configuration. In my setup I have never seen these messages and I needed to access the logs in the container explicitly.

The following is an example setup using the POA Sokol testnet as the Home network, and the Ethereum Kovan testnet as the Foreign network. The instructions for the Bridge UI are identical for an `ERC20-to-ERC20` configuration, but the smart contract deployment steps will vary.

### Dependencies

- [poa-bridge-contracts](https://github.com/poanetwork/poa-bridge-contracts)
- [token-bridge](https://github.com/poanetwork/token-bridge)
- [node.js](https://nodejs.org/en/download/)
- [Nifty Wallet](https://chrome.google.com/webstore/detail/nifty-wallet/jbdaocneiiinmjbjlgalhcelgbejmnid?hl=en) or [MetaMask](https://metamask.io/)

### Example Setup

1. Create an empty folder for setting up your bridge. In this example we call it `sokol-kovan-bridge`.
`mkdir sokol-kovan-bridge && cd sokol-kovan-bridge`  

2. Prepare temporary ETH address(es) for deployment by creating new account(s) in Nifty Wallet or MetaMask. See the [wallet resources](#resources) if you need more information on this step. This account is used:
    * for deploying bridge contracts to both networks
    * as the bridge contracts management wallet
    * as the validator's wallet address(es)

3. Fund the test account(s).

>     * Fund Home accounts (`validators`) using the [POA Sokol Faucet](https://faucet-sokol.herokuapp.com/)

Reviewer: ```suggestion
    * Fund Home account(s) using the [POA Sokol Faucet](https://faucet-sokol.herokuapp.com/)
```

>     * Get free Kovan Coins from the [gitter channel](https://gitter.im/kovan-testnet/faucet) or [Iracus faucet](https://github.com/kovan-testnet/faucet) for Foreign Accounts. Get 5 Keth to 1 acc, and transfer from it to all other wallets.

Reviewer: ```suggestion
    * Get free Kovan Coins from the [gitter channel](https://gitter.im/kovan-testnet/faucet) or [Iracus faucet](https://github.com/kovan-testnet/faucet) for Foreign account(s). Get 5 Keth to 1 acc, and transfer from it to all other wallets if more than one account are used.
```

4. Deploy the Sokol <-> Kovan Bridge contracts.
    * Go to the the `sokol-kovan-bridge` folder created in step 1 and `git clone https://github.com/poanetwork/poa-bridge-contracts`
    * Follow instructions in the [POA Bridge contracts repo](https://github.com/poanetwork/poa-bridge-contracts).
    * Set the parameters in the .env file.
      * `DEPLOYMENT_ACCOUNT_PRIVATE_KEY`: Export the private key from step 2
      * `HOME_RPC_URL`=https://sokol.poa.network
      * Wallet address(es) for bridge contracts management. For testing, you can use the same address for all address values in the file. This includes:
        * `HOME_OWNER_MULTISIG`
        * `HOME_UPGRADEABLE_ADMIN_VALIDATORS`
        * `HOME_UPGRADEABLE_ADMIN_BRIDGE`
        * `FOREIGN_OWNER_MULTISIG`
        * `FOREIGN_UPGRADEABLE_ADMIN_VALIDATORS`
        * `FOREIGN_UPGRADEABLE_ADMIN_BRIDGE`

>         * `VALIDATORS`

Reviewer: ```suggestion
      * Wallet address(es) for validator(s) separated by space. For testing, you can use the same address that was used as the bridge contracts management account. They are specified in the following parameter:
        * `VALIDATORS`
```

>       * `FOREIGN_RPC_URL`=https://kovan.infura.io/mew

Author: not sure if https://kovan.infura.io/mew needs the mew. This was from the previous version. Feel free to eliminate, I believe there are 2 instances in these examples.

Reviewer: it is not an issue - this url is used by MyEthereumWallet application. So, we can leave this as is.

    * When deployment is finished, check that the `bridgeDeploymentResults.json` file exists in the `poa-bridge-contracts/deploy` directory and includes the bridge contract addresses.  

5. Install and run the POA Token Bridge.
    * Got to the `sokol-kovan-bridge` folder and `git clone https://github.com/poanetwork/token-bridge`
    * Follow instructions in the [POA Token Bridge repo](https://github.com/poanetwork/token-bridge).

If successful, you will see bridge processes run when you issue a command. For example, run `npm run watcher:signature-request`

**Example NPM Output:**
```bash
[1539195000507] INFO (watcher-signature-request): Connected to redis
[1539195000545] INFO (watcher-signature-request): Connected to amqp Broker
[1539195006085] INFO (watcher-signature-request): Found 0 UserRequestForSignature events
[1539195011467] INFO (watcher-signature-request): Found 0 UserRequestForSignature events
```

**Example Docker Output:**

**Note:** The output will depend on your Docker configuration. You may need to access the container logs to view.

```bash

> {"level":30,"time":1539366879816,"msg":"Connected to redis","validator":"0xC9f74f16F3743f4370c66DdA728D1751BCBa2616","name":"watcher-signature-request","v":1}

Reviewer: ```suggestion
{"level":30,"time":1539366879816,"msg":"Connected to redis","validator":"0x..........","name":"watcher-signature-request","v":1}
```

> {"level":30,"time":1539366879880,"msg":"Connected to amqp Broker","validator":"0xC9f74f16F3743f4370c66DdA728D1751BCBa2616","name":"watcher-signature-request","v":1}

Reviewer: ```suggestion
{"level":30,"time":1539366879880,"msg":"Connected to amqp Broker","validator":"0x..........","name":"watcher-signature-request","v":1}
```

### Example Setup

1. Create an empty folder for setting up your bridge. In this example we call it `sokol-kovan-bridge`.
`mkdir sokol-kovan-bridge && cd sokol-kovan-bridge`  

> 2. Prepare temporary ETH address(es) for deployment by creating new account(s) in Nifty Wallet or MetaMask. See the [wallet resources](#resources) if you need more information on this step.  

Reviewer: This account will be used:
* for deployment of the bridge contracts to both networks
* as bridge contracts management wallet
* as the validator

3. Fund the test account(s).
    * Fund Home accounts (`Validators`) using the [POA Sokol Faucet](https://faucet-sokol.herokuapp.com/)
    * Get free Kovan Coins from the [gitter channel](https://gitter.im/kovan-testnet/faucet) or [Iracus faucet](https://github.com/kovan-testnet/faucet) for Foreign Accounts. Get 5 Keth to 1 acc, and transfer from it to all other wallets.

4. Deploy the Sokol <-> Kovan Bridge contracts.
    * Go to the the `sokol-kovan-bridge` folder created in step 1 and `git clone https://github.com/poanetwork/poa-bridge-contracts`
    * Follow instructions in the [POA Bridge contracts repo](https://github.com/poanetwork/poa-bridge-contracts).
    * Set the parameters in the .env file.
      * `DEPLOYMENT_ACCOUNT_PRIVATE_KEY`: Export the private key from step 2
      * `HOME_RPC_URL`=https://sokol.poa.network

>       * Wallet address(es) for the validators. For testing, you can use the same address for all address values in the file. This includes:

Reviewer: These are not addresses for the validators. They are for bridge contracts management wallet.

        * `HOME_OWNER_MULTISIG`
        * `HOME_UPGRADEABLE_ADMIN_VALIDATORS`
        * `HOME_UPGRADEABLE_ADMIN_BRIDGE`
        * `FOREIGN_OWNER_MULTISIG`
        * `FOREIGN_UPGRADEABLE_ADMIN_VALIDATORS`
        * `FOREIGN_UPGRADEABLE_ADMIN_BRIDGE`

>       * FOREIGN_RPC_URL=https://kovan.infura.io/mew

Reviewer: let's add the validator configuration since the wallet address funded above will be set up also here.
